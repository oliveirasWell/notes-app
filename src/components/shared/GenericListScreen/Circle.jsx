import React from "react";
import { View, StyleSheet, Pressable } from "react-native";

import { palette } from "@/theme/palette";

const styles = StyleSheet.create({
  circle: {
    justifyContent: "center",
    alignItems: "center",
  },
});

export const CircleCheckBox = ({ active, onPress, size }) => {
  const activeColor = palette.primary[500];
  const inactiveColor = palette.global.white;

  const circleStyle = {
    width: size,
    height: size,
    borderRadius: size / 2,
    backgroundColor: active ? activeColor : inactiveColor,
    borderColor: !active && palette.global.black,
    borderWidth: !active && 1,
  };

  return (
    <Pressable onPress={onPress}>
      <View style={[styles.circle, circleStyle]} />
    </Pressable>
  );
};
