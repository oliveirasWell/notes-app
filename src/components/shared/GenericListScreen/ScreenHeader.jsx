import { useNavigation } from "@react-navigation/native";
import React from "react";
import { StyleSheet, Text, View, Pressable } from "react-native";

import { palette } from "@/theme/palette";
import { spacing } from "@/theme/spacing";

const screenHeaderStyles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: palette.global.white,
    padding: spacing.large,
    flexDirection: "column",
  },
  header: {
    width: "100%",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: palette.primary[100],
    padding: spacing.large,
  },
  headerContainer: { flex: 1, alignItems: "center" },
  headerText: {
    color: palette.primary[500],
    fontWeight: "700",
    fontSize: 20,
    lineHeight: 24,
    fontFamily: "Inter-Bold",
  },
  backButton: { textAlign: "right" },
});

export const ScreenHeader = ({ title, showBackButton }) => {
  const navigation = useNavigation();

  return (
    <View style={screenHeaderStyles.header}>
      {showBackButton && (
        <Pressable onPress={navigation.goBack}>
          <Text style={screenHeaderStyles.backButton}>Back</Text>
        </Pressable>
      )}
      <View style={screenHeaderStyles.headerContainer}>
        <Text style={screenHeaderStyles.headerText}>{title}</Text>
      </View>
    </View>
  );
};
