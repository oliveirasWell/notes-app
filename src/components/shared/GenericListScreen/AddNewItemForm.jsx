import React from "react";
import { StyleSheet, Text, View, Pressable, TextInput } from "react-native";

import { ScreenHeader } from "./ScreenHeader";

import { palette } from "@/theme/palette";
import { spacing } from "@/theme/spacing";

const itemFormStyles = StyleSheet.create({
  inputLabel: {
    color: palette.global.black,
    marginBottom: 5,
    fontFamily: "Inter-Regular",
    fontSize: 12,
    fontWeight: "500",
    lineHeight: 14,
    letterSpacing: 0,
    textAlign: "left",
  },
  inputErrorLabel: {
    color: palette.error[500],
    marginBottom: 5,
    fontFamily: "Inter-Regular",
    fontSize: 12,
    fontWeight: "500",
    lineHeight: 14,
    letterSpacing: 0,
    textAlign: "left",
  },
  inputField: {
    padding: 10,
    borderRadius: 8,
    borderColor: palette.neutral[300],
    borderWidth: 1,
    color: palette.neutral[600],
  },
  label: { paddingTop: spacing.medium, flexDirection: "row" },
  addButtonContainer: {
    marginTop: 5 * spacing.large,
  },
  addButton: {
    width: "100%",
    borderRadius: 30,
    backgroundColor: palette.primary[500],
    paddingHorizontal: 14,
    paddingVertical: 20,
  },
  addButtonText: {
    textAlign: "center",
    color: palette.global.white,
    fontFamily: "Inter-Regular",
    fontSize: 16,
    fontWeight: "700",
    lineHeight: 19,
    letterSpacing: 0,
  },
});

export const AddNewItemForm = ({ onSubmit, formTitle, parentIndex }) => {
  const [value, onChangeText] = React.useState("");
  const onPress = () => {
    onSubmit({ name: value, parentIndex });
    onChangeText("");
  };

  return (
    <View>
      <ScreenHeader title={formTitle} showBackButton={false} />
      <View>
        <View style={itemFormStyles.label}>
          <Text style={itemFormStyles.inputLabel}>Name</Text>
          <Text style={itemFormStyles.inputErrorLabel}>*</Text>
        </View>
        <TextInput
          editable
          numberOfLines={1}
          onChangeText={(text) => onChangeText(text)}
          value={value}
          style={itemFormStyles.inputField}
          placeholder="New Name"
        />
      </View>

      <View style={itemFormStyles.addButtonContainer}>
        <Pressable onPress={onPress} disabled={!value}>
          <View style={itemFormStyles.addButton}>
            <Text style={itemFormStyles.addButtonText}>Add</Text>
          </View>
        </Pressable>
      </View>
    </View>
  );
};
