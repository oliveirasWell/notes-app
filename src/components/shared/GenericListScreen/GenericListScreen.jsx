import React from "react";
import { StyleSheet, Text, FlatList, View, Pressable } from "react-native";

import { AddNewItemForm } from "./AddNewItemForm";
import { ScreenHeader } from "./ScreenHeader";

import { CircleCheckBox } from "@/components/shared/GenericListScreen/Circle";
import { palette } from "@/theme/palette";
import { spacing } from "@/theme/spacing";

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: palette.global.white,
    padding: spacing.large,
    flexDirection: "column",
  },
  listItem: {
    padding: spacing.large,
    flexDirection: "row",
    alignContent: "center",
  },
  listItemContainer: {
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
  },
  noteNameText: {
    color: palette.global.black,
    fontFamily: "Inter-Bold",
    fontSize: 16,
    fontWeight: "700",
    lineHeight: 19,
    letterSpacing: 0,
    alignSelf: "flex-start",
    marginLeft: spacing.medium,
  },
  removeItemText: {
    color: palette.primary[500],
    fontFamily: "Inter-Bold",
    fontSize: 14,
    fontWeight: "700",
    lineHeight: 16.8,
    letterSpacing: 0,
    alignSelf: "flex-end",
    textDecorationLine: "underline",
  },
  separator: {
    backgroundColor: palette.neutral[300],
    width: "100%",
    height: 1,
  },
  flatList: { flex: 1 },
  label: { paddingTop: spacing.medium, flexDirection: "row" },
});

export const GenericListScreen = ({
  list,
  parentIndex = undefined,
  headerTitle,
  removeItem,
  editItem,
  toggleActiveItem,
  showActiveToggle,
  onSubmitNewItem,
  addNewItemTitle,
  showBackButton,
}) => {
  const renderItem = ({ item, index }) => (
    <Pressable
      onPress={() => {
        if (!editItem) {
          return;
        }

        return editItem(index);
      }}
    >
      <View key={index} style={styles.listItem}>
        {showActiveToggle && (
          <CircleCheckBox
            active={item.active}
            onPress={() => toggleActiveItem(item.parentIndex, index)}
            size={30}
          />
        )}
        <View style={styles.listItemContainer}>
          <Text style={styles.noteNameText} numberOfLines={0}>
            {item.name}
          </Text>
          <Pressable onPress={() => removeItem(index, item.parentIndex)}>
            <Text style={styles.removeItemText} numberOfLines={0}>
              Remove item
            </Text>
          </Pressable>
        </View>
      </View>
    </Pressable>
  );

  return (
    <View style={styles.container}>
      <ScreenHeader title={headerTitle} showBackButton={showBackButton} />
      <FlatList
        style={styles.flatList}
        ItemSeparatorComponent={() => <View style={styles.separator} />}
        data={list}
        renderItem={renderItem}
      />
      <AddNewItemForm
        onSubmit={onSubmitNewItem}
        formTitle={addNewItemTitle}
        parentIndex={parentIndex}
      />
    </View>
  );
};
