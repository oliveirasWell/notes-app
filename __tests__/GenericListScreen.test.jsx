import { NavigationContainer } from "@react-navigation/native";
import { render, fireEvent } from "@testing-library/react-native";
import React from "react";

import { GenericListScreen } from "@/components/shared/GenericListScreen/GenericListScreen";

describe("<GenericListScreen />", () => {
  const mockList = [
    { name: "Item 1", active: false },
    { name: "Item 2", active: true },
  ];

  const defaultProps = {
    list: mockList,
    headerTitle: "Test Header",
    removeItem: jest.fn(),
    editItem: jest.fn(),
    toggleActiveItem: jest.fn(),
    showActiveToggle: true,
    onSubmitNewItem: jest.fn(),
    addNewItemTitle: "Add New Item",
    showBackButton: true,
  };

  it("renders correctly", () => {
    const { getByText } = render(
      <NavigationContainer>
        <GenericListScreen {...defaultProps} />
      </NavigationContainer>,
    );
    expect(getByText("Test Header")).toBeTruthy();
    expect(getByText("Item 1")).toBeTruthy();
    expect(getByText("Item 2")).toBeTruthy();
    expect(getByText("Add New Item")).toBeTruthy();
  });

  it("calls editItem on press", () => {
    const { getByText } = render(
      <NavigationContainer>
        <GenericListScreen {...defaultProps} />
      </NavigationContainer>,
    );
    fireEvent.press(getByText("Item 1"));
    expect(defaultProps.editItem).toHaveBeenCalledWith(0);
  });

  it("matches snapshot", () => {
    const tree = render(
      <NavigationContainer>
        <GenericListScreen {...defaultProps} />
      </NavigationContainer>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
