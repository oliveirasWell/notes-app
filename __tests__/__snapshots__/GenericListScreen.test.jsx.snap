// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenericListScreen /> matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": "column",
      "padding": 16,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#D9EFFF",
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          {
            "textAlign": "right",
          }
        }
      >
        Back
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#0C8CE9",
            "fontFamily": "Inter-Bold",
            "fontSize": 20,
            "fontWeight": "700",
            "lineHeight": 24,
          }
        }
      >
        Test Header
      </Text>
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      [
        {
          "active": false,
          "name": "Item 1",
        },
        {
          "active": true,
          "name": "Item 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              {
                "alignContent": "center",
                "flexDirection": "row",
                "padding": 16,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "justifyContent": "center",
                    },
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderColor": "#000000",
                      "borderRadius": 15,
                      "borderWidth": 1,
                      "height": 30,
                      "width": 30,
                    },
                  ]
                }
              />
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                numberOfLines={0}
                style={
                  {
                    "alignSelf": "flex-start",
                    "color": "#000000",
                    "fontFamily": "Inter-Bold",
                    "fontSize": 16,
                    "fontWeight": "700",
                    "letterSpacing": 0,
                    "lineHeight": 19,
                    "marginLeft": 8,
                  }
                }
              >
                Item 1
              </Text>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  numberOfLines={0}
                  style={
                    {
                      "alignSelf": "flex-end",
                      "color": "#0C8CE9",
                      "fontFamily": "Inter-Bold",
                      "fontSize": 14,
                      "fontWeight": "700",
                      "letterSpacing": 0,
                      "lineHeight": 16.8,
                      "textDecorationLine": "underline",
                    }
                  }
                >
                  Remove item
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "#BDC5D0",
              "height": 1,
              "width": "100%",
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              {
                "alignContent": "center",
                "flexDirection": "row",
                "padding": 16,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "justifyContent": "center",
                    },
                    {
                      "backgroundColor": "#0C8CE9",
                      "borderColor": false,
                      "borderRadius": 15,
                      "borderWidth": false,
                      "height": 30,
                      "width": 30,
                    },
                  ]
                }
              />
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                numberOfLines={0}
                style={
                  {
                    "alignSelf": "flex-start",
                    "color": "#000000",
                    "fontFamily": "Inter-Bold",
                    "fontSize": 16,
                    "fontWeight": "700",
                    "letterSpacing": 0,
                    "lineHeight": 19,
                    "marginLeft": 8,
                  }
                }
              >
                Item 2
              </Text>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  numberOfLines={0}
                  style={
                    {
                      "alignSelf": "flex-end",
                      "color": "#0C8CE9",
                      "fontFamily": "Inter-Bold",
                      "fontSize": 14,
                      "fontWeight": "700",
                      "letterSpacing": 0,
                      "lineHeight": 16.8,
                      "textDecorationLine": "underline",
                    }
                  }
                >
                  Remove item
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#D9EFFF",
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 16,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "#0C8CE9",
              "fontFamily": "Inter-Bold",
              "fontSize": 20,
              "fontWeight": "700",
              "lineHeight": 24,
            }
          }
        >
          Add New Item
        </Text>
      </View>
    </View>
    <View>
      <View
        style={
          {
            "flexDirection": "row",
            "paddingTop": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#000000",
              "fontFamily": "Inter-Regular",
              "fontSize": 12,
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 14,
              "marginBottom": 5,
              "textAlign": "left",
            }
          }
        >
          Name
        </Text>
        <Text
          style={
            {
              "color": "#F63737",
              "fontFamily": "Inter-Regular",
              "fontSize": 12,
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 14,
              "marginBottom": 5,
              "textAlign": "left",
            }
          }
        >
          *
        </Text>
      </View>
      <TextInput
        editable={true}
        numberOfLines={1}
        onChangeText={[Function]}
        placeholder="New Name"
        style={
          {
            "borderColor": "#BDC5D0",
            "borderRadius": 8,
            "borderWidth": 1,
            "color": "#666666",
            "padding": 10,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "marginTop": 80,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            {
              "backgroundColor": "#0C8CE9",
              "borderRadius": 30,
              "paddingHorizontal": 14,
              "paddingVertical": 20,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "fontWeight": "700",
                "letterSpacing": 0,
                "lineHeight": 19,
                "textAlign": "center",
              }
            }
          >
            Add
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
